ludwig/utils/strings_utils.py

```diff
@@ -295,65 +376,49 @@ def create_vocabulary(
 
     TODO(Justin): Clean up pad_idx, padding_symbol, unknown_symbol return, as no one seems to be using it.
     """
-    vocab = None
-
     tokenizer = get_tokenizer_from_registry(tokenizer_type)(
         vocab_file=vocab_file,
         pretrained_model_name_or_path=pretrained_model_name_or_path,
         ngram_size=ngram_size,
     )
 
-    # Pre-trained huggingface tokenizer. Use the pre-existing vocabulary and special symbols.
-    if tokenizer_type == "hf_tokenizer":
-        try:
-            vocab = tokenizer.get_vocab()
-            vocab = list(vocab.keys())
-        except NotImplementedError:
-            logger.warning(
-                "HuggingFace tokenizer does not have a get_vocab() method. "
-                + "Using tokenizer.tokenizer.vocab_size and tokenizer.tokenizer._convert_id_to_token "
-                + "to build the vocabulary."
-            )
-            vocab = []
-            for idx in range(tokenizer.tokenizer.vocab_size):
-                vocab.append(tokenizer.tokenizer._convert_id_to_token(idx))
-            vocab += tokenizer.tokenizer.added_tokens_encoder.keys()
-
-        pad_token = tokenizer.get_pad_token()
-        unk_token = tokenizer.get_unk_token()
-
-        if unk_token is None:
-            logger.warning(
-                "No unknown token found in HuggingFace tokenizer. Adding one. "
-                + "NOTE: This will change the vocabulary size and may affect model "
-                + "performance, particularly if the model weights are frozen."
-            )
-            vocab = [unknown_symbol] + vocab
-        else:
-            unknown_symbol = unk_token
-
-        if pad_token is None and add_special_symbols:
-            logger.warning(
-                "No padding token found in HuggingFace tokenizer. Adding one. "
-                + "NOTE: This will change the vocabulary size and may affect model "
-                + "performance, particularly if the model weights are frozen."
-            )
-            vocab = [padding_symbol] + vocab
-        else:
-            padding_symbol = pad_token
-    elif hasattr(tokenizer, "get_vocab"):
-        vocab = tokenizer.get_vocab()
-        vocab = list(vocab.keys())
-    elif vocab_file is not None:
-        vocab = load_vocabulary(vocab_file)
+    # Number of tokens in template.
+    prompt_template_num_tokens = -1
+    if prompt_template:
+        prompt_template_num_tokens = len(tokenizer(prompt_template))
```

Comment:
Just to make sure I understand, this would also count the placeholder tokens like `{object}` or `{pronoun}` in the template as tokens and tokenized them right, for e.g.:

```python
from transformers import AutoTokenizer

template = "The {pronoun} sits on the {object}"

tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
tokens = tokenizer.tokenize(template, add_special_tokens=True)
total_tokens = len(tokens)
```

would give you something like:
`['[CLS]', 'the', '{', 'pro', '##no', '##un', '}', 'sits', 'on', 'the', '{', 'object', '}', '[SEP]']` for the tokens and len as 14 (same as just calling tokenizer(data), but it shows you the actual tokens that will get mapped to token IDs).

I'm wondering whether it makes sense to replace the placeholder tokens with blank spaces before calling `tokenizer(prompt_template)` since we ultimately want the "raw" length of the prompt template since these placeholders will actually get replaced by real data. 

```python
import re
from transformers import AutoTokenizer

template = "The {pronoun} sits on the {object}"

tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")


placeholder_pattern = r'\{[^\}]+\}'
placeholder_tokens = re.findall(placeholder_pattern, template)
for token in placeholder_tokens:
    template = template.replace(token, "")

tokens = tokenizer.tokenize(template, add_special_tokens=True)
total_tokens = len(tokens)
```

which would result in `['[CLS]', 'the', 'sits', 'on', 'the', '[SEP]']` as the tokens and a length of 6. 

I think this is probably a more accurate reflection of the tokens in the prompt template but curious what you think!

Reply:
Thanks for writing out this example.

For longer prompts with single input columns, the token difference feels negligible, however you are totally right that excluding the bracketed elements is more accurate and I've incorporated your suggestion.